### DISCUSSION TITLE
"EVMError: Revert" while running  Unit tests

### DISCUSSION THREAD

**Discussion by Amtaf** [Cyfrin/foundry-full-course-cu#498] [Q&A] [answered]
I have been following the video but I ran into an error while doing unit tests, I have a test file 'DSCEngineTest.t.sol'.While running the function testGetUsdValue(),  I got the error as shown on the screenshot,What could be the cause of that error,i'd be glad to get a solution to it.
![Screenshot from 2023-07-31 16-46-13](https://github.com/Cyfrin/foundry-full-course-f23/assets/9764833/ad4eff0f-b757-4206-9af7-08ce124b217f)

here's my test file 'DSCEngineTest.t.sol' code 
```
//SPDX-License-Identifier: MIT

pragma solidity ^0.8.18;

import {Test} from "forge-std/Test.sol";
import {DeployDSC} from "../../script/DeployDSC.s.sol";
import {DecentralizedStableCoin} from "../../src/DecentralizedStableCoin.sol";
import {DSCEngine} from "../../src/DSCEngine.sol";
import {HelperConfig} from "../../script/HelperConfig.s.sol";

contract DSCEngineTest is Test{
    DeployDSC deployer;
    DecentralizedStableCoin dsc;
    DSCEngine dsce;
    HelperConfig config;
    address ethUsdPriceFeed;
    address weth;
    //address 

    function setUp() public {
        deployer = new DeployDSC();
        (dsc, dsce, config) = deployer.run();
        (ethUsdPriceFeed,, weth,,) = config.activeNetworkConfig();
    }


    /////////////////
    //Price Tests //
    ///////////////
    function testGetUsdValue() public {
        uint256 ethAmount = 15e18;
        uint256 expectedUsd = 30000e18;
        uint256 actualUsd = dsce.getUsdValue(weth, ethAmount);
        assertEq(expectedUsd, actualUsd);

    }

}

```
Here's the link to my Github repo with all the code [DSC](https://github.com/Amtaf/Algorithmic-StableCoin)

**Comment by 0xusmanf** [comment, marked as answer]
Its reverting because in your constructor of `DSCEngine.sol` at line `104` you are not mapping the `s_priceFeeds` correctly. You are doing `s_priceFeeds[tokenAddresses[i] = priceFeedAddress[i]];` which is not assigning the pricefeeds and when you call `getUsdValue` function it tries to fetch price feed of the token and in return gets zero address because there is no price feed stored and when it tries to call `latestRoundData()` by zero address, it reverts. Correct it like this in your constructor `s_priceFeeds[tokenAddresses[i]] = priceFeedAddress[i];`.

**Reply by 0xusmanf** [reply]
After correcting this, the `EVMError: Revert` will go away but your test still won't pass, because you are applying the wrong formula to calculate USD price. Check at line `207`, you need to divide it by `PRECISION` instead of multiplying. Correct it too, it should look like this after correction `return ((uint256(price) * ADDITIONAL_FEED_PRECISION) * amount) / PRECISION;`. Now your test will pass.

**Reply by Amtaf** [reply]
You saved the day, thank you so much, it worked
